subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java

```diff
@@ -93,8 +94,9 @@ public ExceptionPlaceholder(Throwable original, Transformer<ExceptionReplacingOb
             toStringRuntimeExec = failure;
         }
 
-        final List<? extends Throwable> causes = extractCauses(throwable);
-        final List<? extends Throwable> suppressed = extractSuppressed(throwable);
+        // If hasCycle, we will skip the causes and suppressed to prevent infinite looping, after all we handled them the first time already
+        final List<? extends Throwable> causes = extractCauses(throwable, hasCycle);
+        final List<? extends Throwable> suppressed = extractSuppressed(throwable, hasCycle);
```

Comment:
Any reason to pass along `hasCycle` instead of making it clear here (instead of the comment) that we skip the causes/suppressed if we detect a cycle?

Essentially:
```
if(cycle) {
   causes = empty
   suppressed = empty
} else {
   extract()
}
```